### ISSUE TITLE
Invalid vectors and pointer for 0.47.02

### ISSUE THREAD

**Issue by cvuchener** [DFHack/df-structures#346]
I don't know if someone is already working on it, or if they are already tracked. Here are a few invalid vectors and pointers I found on v0.47.02 linux64 using df-structures 424ccedfcde133b295d11a8ff4638cefaf167c6d.

- [x] `unit_soul.performance_skills` (`0x100000001` for my dwarves, not a valid pointer)
- [x] `army_controller.unk_44` (vector pointers are `0xffffffffffffffff`, `0xffffffffffffffff`, `0x0`)
- [x] `world_region.forces` (different values, one example: `0x1400000004`, `0x20000001b`, `0x3c090000000008`)
- [x] `world.features.feature_global_idx`, this one could be a valid vector but not of int32_t (size 18 bytes, capacity 32 bytes).
- [x] `world.features.anon_1`, a vector too but not of pointers (size 36 bytes, capacity 64 bytes)

**Comment by BenLubar** [comment]
I loaded the medium world from https://files.dfhack.org/saves/ in legends mode and here's what `check-structures-sanity` found:

```
uninitialized pointers:

df::global::world.entities.all[#].unknown2.region

bad vectors:

df::global::world.world_data.regions[#].forces
df::global::world.army_controllers.all[#].unk_44

missing classes from structures:

creature_interaction_effect_close_open_woundsst
creature_interaction_effect_cure_infectionst
creature_interaction_effect_heal_nervesst
creature_interaction_effect_heal_tissuesst
creature_interaction_effect_reduce_dizzinessst
creature_interaction_effect_reduce_feverst
creature_interaction_effect_reduce_nauseast
creature_interaction_effect_reduce_painst
creature_interaction_effect_reduce_paralysisst
creature_interaction_effect_reduce_swellingst
creature_interaction_effect_regrow_partsst
creature_interaction_effect_special_attack_interactionst
creature_interaction_effect_stop_bleedingst
interaction_effect_change_item_qualityst
interaction_effect_change_weatherst
interaction_effect_create_itemst
interaction_effect_propel_unitst
interaction_effect_summon_unitst
interaction_source_experimentst
```

If you want to run `check-structures-sanity`, make sure you're on the latest version, because there's a pretty big flaw in how it handles pointer vectors that will give you lots of false positives and probably balloon memory to the point where it crashes on older versions of it.

**Comment by cvuchener** [comment]
#347 fixes world_region.forces

> If you want to run check-structures-sanity, make sure you're on the latest version, because there's a pretty big flaw in how it handles pointer vectors that will give you lots of false positives and probably balloon memory to the point where it crashes on older versions of it.

I am not actually using check-structures-sanity, I am experimenting with reading df data from external applications, and I thought trying some validation was a good test for both my code and df-structures. I did double-check with gui/gm-editor that the vectors/pointers were bad.

I just tried a freshly pulled check-structures-sanity (https://github.com/DFHack/dfhack/commit/40bd55c558ee97598f4c4ca9e7c60915c84d3bc7), I saw the same vectors so no false positives, but memory usage went crazy (14Gb when I killed it).

**Comment by PatrikLundell** [comment]
I've failed to find any case of "unk_44" or the unk_v47_X fields added after it being non zero for the army_controller case in any of the saves I've looked at. Given what @cvuchener has seen, it's reasonable to assume at least two of the 4 new padding fields should go before the vector.

**Comment by cvuchener** [comment]
unk_44 (as a vector) does not exist any more since ab5a7d96f62a80d891ffbbeeebd37c25016f954f.

**Comment by PatrikLundell** [comment]
Thanks. I refreshed my copy with the latest data from github an hour ago, and it claimed to get some stuff, but this wasn't updated. I'll look for another red herring to chase ;)

**Comment by cvuchener** [comment]
On linux32, `ui_sidebar_menus.command_line` contains invalid strings and vectors (dfhack crashes on main menu because of that).

```
(gdb) x/20xw &df::global::ui_sidebar_menus->command_line
0x9717238:	0x00000000	0x0962305c	0x00000000	0x00000000
0x9717248:	0x00000000	0xffffffff	0x00000000	0x00000000
0x9717258:	0x0962305c	0x00000000	0x0000003f	0x00000000
0x9717268:	0x00000000	0x00000000	0x00000000	0x00000000
0x9717278:	0x00000000	0x00000000	0x00000000	0x00000000
```
`0x0962305c` is the empty string (`_ZNSs4_Rep20_S_empty_rep_storageE+12`). Adding 4 bytes before `command_line` could fix it. I need to find a way to do that without breaking 64 bits structures, maybe a int32_t or smaller hidden in the padding of the previous compound. Did anyone test dfhack on win32?

**Comment by cvuchener** [comment]
`ui_sidebar_menus.minimap.tile` is also off by 4 bytes on linux64, so the new int32_t is before that. I don't know what to expect in `ui_sidebar_menus.minimap.data` so I cannot comment on that.

**Comment by cvuchener** [comment]
`world.features` is fixed by #359
